distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Problem_Builder.enso

```diff
@@ -0,0 +1,68 @@
+from Standard.Base import all
+
+from Standard.Base.Error.Problem_Behavior as Problem_Behavior_Module import Problem_Behavior, Report_Warning
+import Standard.Table.Internal.Vector_Builder
+
+from Standard.Table.Error as Error_Module import Missing_Input_Columns, Column_Indexes_Out_Of_Range, No_Output_Columns, Duplicate_Column_Selectors, Input_Indices_Already_Matched, Too_Many_Column_Names_Provided, Duplicate_Output_Column_Names, Invalid_Output_Column_Names
+
+type Problem_Builder
+    type Problem_Builder oob_indices duplicate_column_selectors input_indices_already_matched missing_input_columns other
+
+    report_oob_indices indices =
+        here.append_to_ref this.oob_indices indices
+
+    report_duplicate_column_selectors selectors =
+        here.append_to_ref this.duplicate_column_selectors selectors
+
+    report_input_indices_already_matched indices =
+        here.append_to_ref this.input_indices_already_matched indices
+
+    report_missing_input_columns columns =
+        here.append_to_ref this.missing_input_columns columns
+
+    report_other_warning warning =
+        this.other.append warning
+
+    ## Returns a vector containing all reported problems, aggregated.
+    build_problemset : Vector
+    build_problemset =
+        problems = Vector.new_builder
+        build_vector_and_append ref problem_creator =
+            vec = Ref.get ref . build
+            if vec.not_empty then
+                problems.append (problem_creator vec)
+
+        build_vector_and_append this.oob_indices Column_Indexes_Out_Of_Range
+        build_vector_and_append this.duplicate_column_selectors Duplicate_Column_Selectors
+        build_vector_and_append this.input_indices_already_matched Input_Indices_Already_Matched
+        build_vector_and_append this.missing_input_columns Missing_Input_Columns
+        this.other.to_vector.each problems.append
+
+        problems.to_vector
+
+    ## Attaches gathered warnings to the result.
+       Any errors from the `result` take precedence over the ones owned by this
+       builder.
+    attach_to_result : Problem_Behavior -> Any -> Any
+    attach_to_result problem_behavior result =
+        problem_behavior.attach_problems_after result this.build_problemset
+
+    ## Attaches gathered warnings to the result of the provided computation.
+       If in `Report_Error` mode and there are any problems gathered, the first
+       one will be returned as error without even running the computation.
+    attach_to_computation : Problem_Behavior -> Any -> Any
+    attach_to_computation problem_behavior ~computation =
+        problem_behavior.attach_problems_before this.build_problemset computation
+
+## PRIVATE
+   Creates a new helper object for aggregating problems to report.
+new : Problem_Builder
+new =
+    Problem_Builder (Ref.new Vector_Builder.empty) (Ref.new Vector_Builder.empty) (Ref.new Vector_Builder.empty) (Ref.new Vector_Builder.empty) other=Vector.new_builder
```

Comment:
Why not just use `Vector.builder` for this?

Reply:
Because the problems that come to the `Problem_Builder` are already vectors (for example we can get a vector of out of bounds indices).

So if we were using `Vector.Builder`, we would need to do: `vec.each builder.append`. Whereas now we can do `builder ++= vec` (modulo the `Ref` handling). The `++` operation is not performing any copying until `build_problemset` is called. The Vector

Reply:
Feel like we could use a `concat` on the `Vector.builder` doing an `Array.copy` which would be pretty quick too.

Not a big issue and maybe one for tidying up `Vector.builder` vs `Vector_builder` in the future.

Reply:
Yeah, we could add an `append_many` or something to Vector.builder, true.

I wouldn't merge `Vector.Builder` with `Vector_Builder` as they server completely different purposes (one is a mutable data structure and the other is immutable), but we should probably figure out better names for the two - maybe I can call the latter `Immutable_Vector_Builder`? Or since Enso is rather meant to be immutable-by-default, we could rename the first to `Vector.Mutable_Builder`? What do you think?